### PR TITLE
#68 - Improve exponential remainder estimation

### LINKED ISSUE #68: Improve exponential remainder estimation
Implement Proposition 1 in https://mediatum.ub.tum.de/doc/1287218/1287218.pdf

See also https://ieeexplore.ieee.org/document/1446499/

### PR DESCRIPTION
**Pull request by mforets**
Closes #68.

### CODE REVIEW

File: src/exponential.jl [hunk 1]
```diff
@@ -97,7 +97,40 @@ function expm_overapproximation(A::IntervalMatrix{T, Interval{T}}, t, p) where {
...
+    for i in 1:p
+        i! *= i
+        tⁱ *= t
+        Q = Q + Mⁱ * tⁱ/i!
+        Mⁱ = Mⁱ * M
+    end
+    Y  = M - Q
+    Γ = IntervalMatrix(fill(zero(T)±one(T), (n, n)))
+    E = Γ * Y
+    return IntervalMatrix(E) # TODO why do we need the IntervalMatrix?
```
Comment (schillic) [inline src/exponential.jl]: > why do we need the IntervalMatrix?

In the past this happened to me and I found that at some point an `IntervalMatrix` used an operation that is not properly wrapped as an `IntervalMatrix` again. I guess if you step through the algorithm line by line and print the result, you should find the place where you change to an `Array{Interval, 2}`. (However, as a word of warning, sometimes an `IntervalMatrix` is also printed as an `Array{Interval, 2}`. So always be sure to print `typeof(·)`.)

**Comment by schillic** [comment]
Comparing the diameter of the remainder (`diam(E)`) for the transmission-line model with two nodes:
```julia
# master
[63760.82073750474 63760.82073750474 63760.82073750474 63760.82073750474;
 63760.82073750474 63760.82073750474 63760.82073750474 63760.82073750474;
 63760.82073750474 63760.82073750474 63760.82073750474 63760.82073750474;
 63760.82073750474 63760.82073750474 63760.82073750474 63760.82073750474]

# this branch
[0.399656359315619 0.24413731567321428 12.302865896327024 21.43132422399167;
 0.399656359315619 0.24413731567321428 12.302865896327024 21.43132422399167;
 0.399656359315619 0.24413731567321428 12.302865896327024 21.43132422399167;
 0.399656359315619 0.24413731567321428 12.302865896327024 21.43132422399167]
```

File: src/exponential.jl [hunk 2]
```diff
@@ -97,7 +97,40 @@ function expm_overapproximation(A::IntervalMatrix{T, Interval{T}}, t, p) where {
...
 
+# Implementation of Prop. 1 in Althoff, Matthias, Bruce H. Krogh, and Olaf Stursberg.
+# "Analyzing reachability of linear dynamic systems with parametric uncertainties."
+# Modeling, Design, and Simulation of Systems with Uncertainties. Springer, Berlin, Heidelberg, 2011. 69-94.
 function _expm_remainder(A::IntervalMatrix{T}, t, p; n=checksquare(A)) where {T}
+    C = max.(abs.(inf.(A)), abs.(sup.(A)))
+    M = exp(C*t)
+
+    # compute Q = I + Mt + M^2/2! + ... + M^p/p!
+    # where M = exp(abs(A*t)) and abs is taken component-wise
```
Comment (schillic) [inline src/exponential.jl]: The paper uses `C` and not `M` in the sum.

File: src/exponential.jl (hunk 1 above)
Comment (schillic) [reply src/exponential.jl]: Forget what I said. In this case you do `Γ::IntervalMatrix * Y::Matrix`, which we do not define, so I guess the Julia default of `*` interprets `Γ` as a matrix. The proper solution is to define `*(::IntervalMatrix, ::AbstractMatrix)`.

File: src/exponential.jl (hunk 2 above)
Comment (mforets) [reply src/exponential.jl]: Yes, sure, it was a silly mistake. Thanks for finding it.

File: src/exponential.jl [hunk 3]
```diff
@@ -97,7 +97,35 @@ function expm_overapproximation(A::IntervalMatrix{T, Interval{T}}, t, p) where {
     return res
 end
 
+# Implementation of Prop. 1 in Althoff, Matthias, Bruce H. Krogh, and Olaf Stursberg.
```
Comment (schillic) [inline src/exponential.jl]:
```suggestion
# Implementation of Prop. 1 in Matthias Althoff, Bruce H. Krogh, and Olaf Stursberg.
```

File: src/exponential.jl [hunk 4]
```diff
@@ -97,7 +97,35 @@ function expm_overapproximation(A::IntervalMatrix{T, Interval{T}}, t, p) where {
...
+    end
+    M = exp(C*t)
+    Y  = M - Q
+    Γ = IntervalMatrix(fill(zero(T)±one(T), (n, n)))
+    E = Γ * Y
+    return IntervalMatrix(E) # See #73
+end
+
+# Estimates the sum of the series in the matrix exponential. See Theorem 1
+# in [1] Althoff, Matthias, Olaf Stursberg, and Martin Buss.
```
Comment (schillic) [inline src/exponential.jl]:
```suggestion
# in [1] Matthias Althoff, Olaf Stursberg, and Martin Buss.
```

Comment (mforets) [reply src/exponential.jl]: I copied the reference from google scholar. There are different citation styles which one can choose, and I most frequently use [MLA](https://www.library.cornell.edu/sites/default/files/mla_style_revised.pdf), which happens to begin with the the last name of the first author and then use surname-last name of the other authors.

I'm fine to change to another standard style (APA, ISO690, MLA), i have no particular preference.

Comment (schillic) [reply src/exponential.jl]: Okay, we can keep it.

**Pull request merged**